File: Data Structures & Algorithms/string-encode-and-decode/submission-3.py
class Solution:

    def encode(self, strs: List[str]) -> str:
        if not len(strs):
            return str()
        res = ""
        for string in strs:
            temp = ""
            for c in string:
                temp += str(ord(c)) + ","
            temp = temp[:-1]
            res += temp + "+"
        return res

    def decode(self, s: str) -> List[str]:
        if not len(s):
            return []
        s = s[:-1]
        res = s.split('+')
        list_str = []
        for string in res:
            list_str.append('' if string.split(',') == [''] else ''.join([chr(int(x)) for x in string.split(',')]))
        return list_str

File: Data Structures & Algorithms/buy-and-sell-crypto-with-cooldown/submission-3.cpp
class Solution {
public:
    int maxProfit(vector<int>& prices) {
        int n = prices.size();
        if(n == 1)
            return 0;
        vector<int>dp(n,0);
        dp[1] = max(dp[0], prices[1] - prices[0]);
        int res = max(dp[0], dp[1]);
        for(int i = 2; i < n; i++){
            dp[i] = dp[i-1];
            dp[i] = max(dp[i], max(prices[i] - prices[1], prices[i] - prices[0]));
            res = max(res, dp[i]);
            for(int j = 2; j <= i; j++){
                dp[i] = max(dp[i], prices[i] - prices[j] + dp[j-2]);
                res = max(res, dp[i]);
            }
        }
        return res;
    }
};
